docs/merkletree/treetex/main.go

```diff
@@ -0,0 +1,236 @@
+)
+
+const (
+	preamble = `
+% Hash-tree
+% Author: treetex
+\documentclass[convert]{standalone}
+\usepackage[dvipsnames]{xcolor}
+\usepackage{forest}
+
+
+\begin{document}
+
+% Change colours here:
+\definecolor{inclusion}{rgb}{1,0.5,0.5}
+\definecolor{inclusion_ephemeral}{rgb}{1,0.7,0.7}
+\definecolor{perfect}{rgb}{1,0.9,0.5}
+\definecolor{target}{rgb}{0.5,0.5,0.9}
+\definecolor{target_path}{rgb}{0.7,0.7,0.9}
+
+\begin{forest}
+`
+
+	postfix = `\end{forest}
+\end{document}
+`
+	// maxLen is a suitably large maximum nodeID length for storage.NodeID.
+	maxLen = 64
+)
+
+var (
+	treeSize  = flag.Int64("tree_size", 23, "Size of tree to produce")
+	inclusion = flag.Int64("inclusion", -1, "Leaf index to show inclusion proof")
+
+	// nInfo holds nodeInfo data for the tree.
+	nInfo = make(map[string]nodeInfo)
+)
+
+// nodeInfo represents the style to be applied to a tree node.
+type nodeInfo struct {
+	incProof    bool
+	incPath     bool
+	target      bool
+	perfectRoot bool
+	ephemeral   bool
+	leaf        bool
+}
+
+// String returns a string containing Forest attributes suitable for
+// rendering the node, given its type.
+func (n nodeInfo) String() string {
+	attr := make([]string, 0, 4)
+
+	// Figure out which colour to fill with:
+	fill := "white"
+	if n.perfectRoot {
+		attr = append(attr, "line width=4pt")
+	}
+	if n.incProof {
+		fill = "inclusion"
+		if n.ephemeral {
+			fill = "inclusion_ephemeral"
+			attr = append(attr, "draw, dotted")
+		}
+	}
+	if n.target {
+		fill = "target"
+	}
+	if n.incPath {
+		fill = "target_path"
+	}
+	attr = append(attr, "fill="+fill)
+
+	if !n.ephemeral {
+		attr = append(attr, "draw")
+	}
+	if !n.leaf {
+		attr = append(attr, "circle, minimum size=3em")
+	} else {
+		attr = append(attr, "minimum size=1.5em")
+	}
+	return strings.Join(attr, ", ")
+}
+
+// modifyNodeInfo applies f to the nodeInfo associated with node k.
+func modifyNodeInfo(k string, f func(*nodeInfo)) {
+	n, ok := nInfo[k]
+	if !ok {
+		n = nodeInfo{}
+	}
+	f(&n)
+	nInfo[k] = n
+}
+
+// perfect renders a perfect subtree.
+func perfect(prefix string, height, tier, index int64) {
+	perfectInner(prefix, height, tier, index, true)
+}
+
+// drawLeaf emits TeX code to render a leaf.
+func drawLeaf(prefix string, index int64) {
+	a := nInfo[nodeKey(0, index)]
+	fmt.Printf("%s [%d, %s, tier=leaf]\n", prefix, index, a.String())
+}
+
+// openInnerNode renders tex code to open an internal node.
+// The caller may emit any number of child nodes before calling the returned
+// func to clode the node.
+// returns a func to be called to close the node.
+//
+func openInnerNode(prefix string, height, index, tier int64) func() {
+	attr := nInfo[nodeKey(height, index)].String()
+	fmt.Printf("%s [%d.%d, %s, tier=%d\n", prefix, height, index, attr, tier)
+	return func() { fmt.Printf("%s ]\n", prefix) }
+}
+
+// perfectInner renders the nodes of a perfect internal subtree.
+func perfectInner(prefix string, height, tier, index int64, top bool) {
```

Comment:
Since `height` and `tier` in a perfect subtree are basically the same thing (note that they are always decreased simultaneously in lines 162-163), then I suggest having only one parameter.